templates/android/library/src/main/java/io/appwrite/services/Realtime.kt.twig

```diff
@@ -0,0 +1,165 @@
+package {{ sdk.namespace | caseDot }}.services
+
+import com.google.gson.Gson
+import com.google.gson.JsonParseException
+import com.google.gson.JsonSyntaxException
+import {{ sdk.namespace | caseDot }}.Client
+import {{ sdk.namespace | caseDot }}.extensions.forEachAsync
+import {{ sdk.namespace | caseDot }}.models.RealtimeCodes
+import {{ sdk.namespace | caseDot }}.models.RealtimeError
+import {{ sdk.namespace | caseDot }}.models.RealtimeMessage
+import {{ sdk.namespace | caseDot }}.models.RealtimeSubscription
+import kotlinx.coroutines.*
+import kotlinx.coroutines.Dispatchers.IO
+import okhttp3.*
+import okhttp3.internal.concurrent.TaskRunner
+import okhttp3.internal.ws.RealWebSocket
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+class Realtime(client: Client) : Service(client), CoroutineScope {
+
+    private val job = Job()
+
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.Main + job
+
+    private companion object {
+        private const val DEBOUNCE_MILLIS = 1L
+
+        private var socket: RealWebSocket? = null
+        private var channelCallbacks = mutableMapOf<String, MutableCollection<(Any) -> Unit>>()
+        private var errorCallbacks = mutableSetOf<(RealtimeError) -> Unit>()
+
+        private var subCallDepth = 0
+    }
+
+    private fun createSocket() {
+        val queryParamBuilder = StringBuilder()
+            .append("project=${client.config["project"]}")
+
+        channelCallbacks.keys.forEach {
+            queryParamBuilder
+                .append("&channels[]=$it")
+        }
+
+        val request = Request.Builder()
+            .url("${client.endPointRealtime}/realtime?$queryParamBuilder")
+            .build()
+
+        if (socket != null) {
+            closeSocket()
+        }
+
+        socket = RealWebSocket(
+            taskRunner = TaskRunner.INSTANCE,
+            originalRequest = request,
+            listener = {{ spec.title | caseUcfirst }}WebSocketListener(),
+            random = Random(),
+            pingIntervalMillis = client.http.pingIntervalMillis.toLong(),
+            extensions = null,
+            minimumDeflateSize = client.http.minWebSocketMessageToCompress
+        )
+
```

Comment:
We could throw an error here if the socket connection was not created

Reply:
Or does the `RealWebSocket()` constructor already throw an exception

Reply:
The constructor throws if the request method is not `GET`; anything thrown after that will be routed to `onFailure` of the `AppwriteWebsocketListener`. I've changed to a non-null assertion on the `socket` for the `connect` call so any case where instantiation fails but execution continues will throw as well

Reply:
Okay sounds good :+1: